### DISCUSSION TITLE
Use of {crew} as an asynchronous backend for Shiny

### DISCUSSION THREAD

**Discussion by shikokuchuo** [wlandau/crew#69] [General]
Congratulations again on formulating the integration with Shiny! I was playing around with your vignette example and thought you might like to incorporate the following suggestions:

1. Use [5] daemons. It is much easier to see the async in action by clicking say 10 times in succession, then seeing 5 done at a time.
2. Slightly more concise formulation for `status_message()`:

```r
status_message <- function(tasks) {
  if (tasks == 0L) {
    "All tasks completed."
  } else {
    sprintf("%d task%s in progress at %s", tasks, if (tasks > 1L) "s" else "", format.POSIXct(Sys.time()))
  }
}
```

**Comment by wlandau** [comment]
Thanks for the helpful suggestions. I implemented tweaks along those lines in 5b803956ac205bdfb546217881938c1c5c7aaeb9 and 02b20a632bd92fb90006034f8a3af93adc0d492b.

**Comment by shikokuchuo** [comment]
I took a deeper look into the Shiny reactive machinery, and the conclusion (for now) is that polling is the only way to check for resolution of a 'mirai'. This is exemplified by the built-in `reactiveFileReader()`, which is polling - there is no mechanism to automatically update if a file or non-reactive variable changes. There would have to be a function call to invalidate the Shiny context (or equivalent) at least, and this is not possible from another thread; alternatively modifying something in one of the nested environments used by Shiny would also be potentially unsafe.

In light of this, I do have one suggestion that I'll put in a PR for ease of exposition - feel free to take what you want, you don't need to merge it. Just to only poll when there are outstanding tasks.
